src/builtins/path.cpp

```diff
@@ -0,0 +1,892 @@
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto pos = find_extension(*arg);
+
+        wcstring ext;
+        if (!pos) {
+            ext = *arg;
+        } else {
+            ext = arg->substr(0, *pos);
+        }
+
+        // Only add on the extension "." if we have something.
+        // That way specifying an empty extension strips it.
+        if (*opts.arg1) {
+            if (opts.arg1[0] != L'.') {
+                ext.push_back(L'.');
+            }
+            ext.append(opts.arg1);
+        }
+        path_out(streams, opts, ext);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_resolve(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto real = wrealpath(*arg);
+
+        if (!real) {
+            // The path doesn't exist, so we go up until we find
+            // something that does.
+            wcstring next = *arg;
+            // First add $PWD if we're relative
+            if (!next.empty() && next[0] != L'/') {
+                next = wgetcwd() + L"/" + next;
+            }
+            auto rest = wbasename(next);
+            while(!next.empty() && next != L"/") {
+                next = wdirname(next);
+                real = wrealpath(next);
+                if (real) {
+                    real->push_back(L'/');
+                    real->append(rest);
+                    real = normalize_path(*real, false);
+                    break;
+                }
+                rest = wbasename(next) + L'/' + rest;
+            }
+            if (!real) {
+                continue;
+            }
+        }
+
+        // Normalize the path so "../" components are eliminated even after
+        // nonexistent or non-directory components.
+        // Otherwise `path resolve foo/../` will be `$PWD/foo/../` if foo is a file.
+        real = normalize_path(*real, false);
+
+        // Return 0 if we found a realpath.
+        if (opts.quiet) {
+            return STATUS_CMD_OK;
+        }
+        path_out(streams, opts, *real);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_sort(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    opts.invert_valid = true;
+    opts.what_valid = true;
+    opts.unique_valid = true;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    auto func = +[] (const wcstring &x) {
```

Comment:
LOL the `+ lambda` trick, first time encountering this 👏

Reply:
Yeah, it's a workaround because lambda types are unique:

```c++
auto t = []{};
auto t2 = []{};
```

Here, t and t2 have different types. So reassigning breaks - `t = t2` complains about incompatible types. The `+` coerces it to a function pointer so it works (possibly at some slight cost, I dunno). It could also be possible to use std::function instead but that requires us to spell the type again, AFAICT.

See for example https://stackoverflow.com/questions/18889028/a-positive-lambda-what-sorcery-is-this.